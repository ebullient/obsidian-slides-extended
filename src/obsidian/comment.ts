import { Properties } from "./transformers";

export class Comment extends Properties {
    type: string;

    constructor(type: string, attributes: Map<string, string>) {
        attributes.set("onTarget", type);
        super(attributes);
        this.type = type;
    }

    public static of(
        type: string,
        style: string[] = [],
        clazz: string[] = [],
        attributes: Map<string, string> = new Map<string, string>(),
    ): Comment {
        if (clazz && clazz.length > 0) {
            const classValue = attributes.get("class");

            if (classValue) {
                const split = classValue.split(" ");
                split.push(...clazz);
                attributes.set("class", split.join(" "));
            } else {
                attributes.set("class", clazz.join(" "));
            }
        }

        if (style && style.length > 0) {
            const styleValue = attributes.get("style");

            if (styleValue) {
                const split = styleValue
                    .split(";")
                    .map((value) => value.trim());
                split.push(...style);
                attributes.set("style", split.join("; "));
            } else {
                attributes.set("style", style.join("; "));
            }
        }

        return new Comment(type, attributes);
    }
}

export class CommentParser {
    private readCommentRegex = /<!--.*?-->/s;
    private parseRegex = /<!--\s*(?:\.)?(element|slide):?\s*(.*)-->/s;
    private parsePropertiesRegex =
        /([^=]*)\s*=\s*"([^"]*)"\s*|([^=]*)\s*=\s*'([^']*)'\s*|([^ ]*)\s*/g;

    commentToString(comment: Comment): string {
        return `<!-- .${comment.type}: ${this.buildAttributes(comment)} -->`;
    }

    buildAttributes(comment: Comment): string {
        const styles = comment.getStyles();
        const classes = comment.getClasses();

        const stylesString = styles.length > 0 ? `style="${styles}" ` : "";
        const classesString = classes.length > 0 ? `class="${classes}" ` : "";

        return `${stylesString}${classesString}${comment.getAttributes()}`.trim();
    }

    parseLine(line: string): Comment {
        return this.lineHasComment(line)
            ? this.parseComment(this.readCommentStringFromLine(line))
            : null;
    }

    parseComment(comment: string): Comment {
        try {
            const [, type, properties] = this.parseRegex.exec(comment);
            const attributes = this.parseAttributes(properties);

            return new Comment(type, attributes);
        } catch (_ex) {
            console.error(`ERROR: Cannot parse comment: ${comment}`);
            return null;
        }
    }

    buildComment(
        type: string,
        style: string[] = [],
        clazz: string[] = [],
        attributes: Map<string, string> = new Map<string, string>(),
    ): Comment {
        return Comment.of(type, style, clazz, attributes);
    }

    lineHasComment(line: string): boolean {
        return this.parseRegex.test(line);
    }

    replace(line: string, comment: Comment): string {
        return line.replace(
            this.readCommentRegex,
            this.commentToString(comment),
        );
    }

    private readCommentStringFromLine(line: string): string {
        return this.readCommentRegex.exec(line)?.[0] ?? "";
    }

    private parseAttributes(properties: string): Map<string, string> {
        const attributes = new Map<string, string>();

        while (true) {
            const m = this.parsePropertiesRegex.exec(properties);
            if (m == null) {
                break;
            }
            if (m.index === this.parsePropertiesRegex.lastIndex) {
                this.parsePropertiesRegex.lastIndex++;
            }

            let key: string;
            let value: string;

            m.forEach((match, groupIndex) => {
                if (groupIndex === 1 || groupIndex === 3) {
                    key = match;
                }
                if (groupIndex === 2 || groupIndex === 4) {
                    value = match;
                    attributes.set(key, value);
                }
                if (groupIndex === 5) {
                    if (match) {
                        attributes.set(match, "true");
                    }
                }
            });
        }

        attributes.delete(undefined);
        return attributes;
    }
}
