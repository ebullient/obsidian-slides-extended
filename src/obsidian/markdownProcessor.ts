import type { Options, Processor } from "../@types";
import { YamlStore } from "../yaml/yamlStore";
import type { ObsidianUtils } from "./obsidianUtils";
import { AutoClosingProcessor } from "./processors/autoClosingProcessor";
import { BlockProcessor } from "./processors/blockProcessor";
import { CalloutProcessor } from "./processors/calloutProcessor";
import { ChartProcessor } from "./processors/chartProcessor";
import { CommentProcessor } from "./processors/commentProcessor";
import { DebugViewProcessor } from "./processors/debugViewProcessor";
import { DefaultBackgroundProcessor } from "./processors/defaultBackgroundProcessor";
import { DropProcessor } from "./processors/dropProcessor";
import { EmojiProcessor } from "./processors/emojiProcessor";
import { ExcalidrawProcessor } from "./processors/excalidrawProcessor";
import { FootnoteProcessor } from "./processors/footNoteProcessor";
import { FormatProcessor } from "./processors/formatProcessor";
import { FragmentProcessor } from "./processors/fragmentProcessor";
import { GridProcessor } from "./processors/gridProcessor";
import { IconsProcessor } from "./processors/iconsProcessor";
import { InternalLinkProcessor } from "./processors/internalLinkProcessor";
import { LatexProcessor } from "./processors/latexProcessor";
import { MediaProcessor } from "./processors/mediaProcessor";
import { MermaidProcessor } from "./processors/mermaidProcessor";
import { MultipleFileProcessor } from "./processors/multipleFileProcessor";
import { ReferenceProcessor } from "./processors/referenceProcessor";
import { SkipSlideProcessor } from "./processors/skipSlideProcessor";
import { TemplateProcessor } from "./processors/templateProcessor";

interface ProcessStep {
    name: string;
    processor: Processor;
}

export class MarkdownProcessor {
    private multipleFileProcessor: MultipleFileProcessor;
    private blockProcessor: BlockProcessor;
    private mediaProcessor: MediaProcessor;
    private internalLinkProcessor: InternalLinkProcessor;
    private footnoteProcessor: FootnoteProcessor;
    private latexProcessor: LatexProcessor;
    private formatProcessor: FormatProcessor;
    private excalidrawProcessor: ExcalidrawProcessor;
    private mermaidProcessor: MermaidProcessor;
    private fragmentProcessor: FragmentProcessor;
    private gridProcessor: GridProcessor;
    private commentProcessor: CommentProcessor;
    private dropProcessor: DropProcessor;
    private autoClosingProcessor: AutoClosingProcessor;
    private emojiProcessor: EmojiProcessor;
    private iconsProcessor: IconsProcessor;
    private debugViewProcessor: DebugViewProcessor;
    private calloutProcessor: CalloutProcessor;
    private templateProcessor: TemplateProcessor;
    private chartProcessor: ChartProcessor;
    private defaultBackgroundProcessor: DefaultBackgroundProcessor;
    private referenceProcessor: ReferenceProcessor;
    private skipSlideProcessor: SkipSlideProcessor;
    private stripLatexBackTicks: Processor;

    constructor(utils: ObsidianUtils) {
        this.multipleFileProcessor = new MultipleFileProcessor(utils);
        this.blockProcessor = new BlockProcessor();
        this.mediaProcessor = new MediaProcessor(utils);
        this.internalLinkProcessor = new InternalLinkProcessor(utils);
        this.footnoteProcessor = new FootnoteProcessor();
        this.latexProcessor = new LatexProcessor();
        this.formatProcessor = new FormatProcessor();
        this.excalidrawProcessor = new ExcalidrawProcessor(utils);
        this.mermaidProcessor = new MermaidProcessor();
        this.fragmentProcessor = new FragmentProcessor();
        this.gridProcessor = new GridProcessor();
        this.commentProcessor = new CommentProcessor();
        this.dropProcessor = new DropProcessor();
        this.autoClosingProcessor = new AutoClosingProcessor();
        this.emojiProcessor = new EmojiProcessor();
        this.iconsProcessor = new IconsProcessor();
        this.debugViewProcessor = new DebugViewProcessor();
        this.calloutProcessor = new CalloutProcessor();
        this.templateProcessor = new TemplateProcessor(utils);
        this.chartProcessor = new ChartProcessor();
        this.defaultBackgroundProcessor = new DefaultBackgroundProcessor();
        this.referenceProcessor = new ReferenceProcessor();
        this.skipSlideProcessor = new SkipSlideProcessor();
        this.stripLatexBackTicks = {
            process: (markdown: string, _: Options) => {
                return markdown.replaceAll("%`%", "");
            },
        };
    }

    process(markdown: string, options: Options) {
        YamlStore.getInstance().options = options;

        let processedMarkdown = this.trimEnding(markdown, options);
        if (options.log) {
            this.log("begin", markdown, processedMarkdown);
        }

        // First phase: Template processing
        processedMarkdown = this.processTemplates(processedMarkdown, options);

        // Second phase: Core processors that modify slide structure
        processedMarkdown = this.processSlideStructure(
            processedMarkdown,
            options,
        );

        // Third phase: Content processors
        processedMarkdown = this.processContent(processedMarkdown, options);

        if (options.log) {
            this.log("end", markdown, processedMarkdown);
        }
        return processedMarkdown;
    }

    private processWithLog(
        before: string,
        options: Options,
        step: ProcessStep,
    ): string {
        const after = step.processor.process(before, options);
        if (options.log) {
            this.log(step.name, before, after);
        }
        return after;
    }

    private processTemplates(markdown: string, options: Options): string {
        // Process multi-file includes and templates
        // Trim trailing separators
        let before = markdown;
        let after: string;

        let circuitCounter = 0;
        while (before !== after) {
            circuitCounter++;
            if (after) {
                before = after;
            }

            // Process multiple file includes first
            after = this.processWithLog(before, options, {
                name: "multipleFileProcessor",
                processor: this.multipleFileProcessor,
            });

            // Then process templates
            after = this.processWithLog(after, options, {
                name: "templateProcessor",
                processor: this.templateProcessor,
            });

            // Remove default templates after first pass to prevent re-application
            options.defaultTemplate = null;

            // Circuit breaker to prevent infinite loops
            if (circuitCounter > 9) {
                console.warn(
                    "WARNING: Circuit in template hierarchy detected!",
                );
                break;
            }
        }
        if (options.log) {
            this.log("merge & template", markdown, after);
        }
        return after || before;
    }

    private processSlideStructure(markdown: string, options: Options): string {
        // Process slide structural elements in a specific order
        return [
            // Skip slides marked to be hidden
            {
                name: "skipSlideProcessor",
                processor: this.skipSlideProcessor,
            },
            // Add debug grid if enabled
            {
                name: "debugViewProcessor",
                processor: this.debugViewProcessor,
            },
            // Auto-close self-closing tags
            {
                name: "autoClosingProcessor",
                processor: this.autoClosingProcessor,
            },
            // Apply default backgrounds
            {
                name: "defaultBackgroundProcessor",
                processor: this.defaultBackgroundProcessor,
            },
        ].reduce(
            (md, step) => this.processWithLog(md, options, step),
            markdown,
        );
    }

    private processContent(markdown: string, options: Options): string {
        // Process content elements in a specific order
        return [
            // Process LaTeX content
            { name: "latexProcessor", processor: this.latexProcessor },
            // Process callouts
            {
                name: "calloutProcessor",
                processor: this.calloutProcessor,
            },
            // Convert emoji shortcodes
            { name: "emojiProcessor", processor: this.emojiProcessor },
            // Process icon shortcodes
            { name: "iconsProcessor", processor: this.iconsProcessor },
            // Process formatting
            { name: "formatProcessor", processor: this.formatProcessor },
            // Convert mermaid code blocks
            {
                name: "mermaidProcessor",
                processor: this.mermaidProcessor,
            },
            // Process block syntax
            { name: "blockProcessor", processor: this.blockProcessor },
            // Process footnotes
            {
                name: "footnoteProcessor",
                processor: this.footnoteProcessor,
            },
            // Process excalidraw embeds
            {
                name: "excalidrawProcessor",
                processor: this.excalidrawProcessor,
            },
            // Process media embeds
            { name: "mediaProcessor", processor: this.mediaProcessor },
            // Process internal links
            {
                name: "internalLinkProcessor",
                processor: this.internalLinkProcessor,
            },
            // Process reference blockrefs
            {
                name: "referenceProcessor",
                processor: this.referenceProcessor,
            },
            // Process fragments
            {
                name: "fragmentProcessor",
                processor: this.fragmentProcessor,
            },
            // Process drop layouts
            { name: "dropProcessor", processor: this.dropProcessor },
            // Process grid layouts
            { name: "gridProcessor", processor: this.gridProcessor },
            // Process slide comments
            {
                name: "commentProcessor",
                processor: this.commentProcessor,
            },
            // Process charts
            { name: "chartProcessor", processor: this.chartProcessor },
            {
                name: "stripLatexBackTicks",
                processor: this.stripLatexBackTicks,
            },
        ].reduce(
            (md, step) => this.processWithLog(md, options, step),
            markdown,
        );
    }

    postProcess = (element: HTMLElement) => {
        const paragraphs = element.findAll("p");
        for (const paragraph of paragraphs) {
            if (paragraph.innerText.startsWith(":::")) {
                paragraph.remove();
            }
        }
    };

    trimEnding(markdown: string, options: Options): string {
        const input = `${markdown}\n`;

        let m = new RegExp(options.separator, "gmi").exec(input);
        if (m !== null) {
            const [match] = m;

            if (input.endsWith(match)) {
                return input.substring(0, input.lastIndexOf(match));
            }
        }

        m = new RegExp(options.verticalSeparator, "gmi").exec(input);
        if (m !== null) {
            const [match] = m;

            if (input.endsWith(match)) {
                return input.substring(0, input.lastIndexOf(match));
            }
        }

        return markdown;
    }

    log(name: string, before: string, after: string) {
        if (before !== after) {
            console.debug(name, JSON.stringify({ before, after }));
        }
    }
}
