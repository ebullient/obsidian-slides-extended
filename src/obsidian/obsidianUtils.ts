import path from "node:path";
import { readFileSync } from "fs-extra";
import {
    type App,
    FileSystemAdapter,
    Platform,
    resolveSubpath,
    type TFile,
} from "obsidian";
import type {
    MediaCollector,
    Options,
    SlidesExtendedSettings,
} from "../@types";
import { DISABLED_IMAGE_COLLECTOR } from "../slidesExtended-constants";
import { MarkdownProcessor } from "./markdownProcessor";

let instance: MediaCollector = DISABLED_IMAGE_COLLECTOR;
export function getMediaCollector(): MediaCollector {
    return instance;
}

function setMediaCollector(newInstance: MediaCollector) {
    instance = newInstance;
}

export function processBySlide(
    markdown: string,
    options: Options,
    processOneSlide: (slide: string) => string,
): string {
    return markdown
        .split(new RegExp(options.separator, "gmi"))
        .map((slidegroup) => {
            return slidegroup
                .split(new RegExp(options.verticalSeparator, "gmi"))
                .map((slide) => {
                    return processOneSlide(slide);
                })
                .join(options.verticalSeparator);
        })
        .join(options.separator);
}

export function skipMathCodeBlocks(
    markdown: string,
    processRemainingText: (markdown: string) => string,
): string {
    return skipCodeBlocks(markdown, (md) => {
        const parts = md.split(/(%`%[\s\S]+?%`%)/g);
        for (let i = 0; i < parts.length; i++) {
            if (parts[i].startsWith("%`%")) {
                // Don't modify protected math segments at all
            } else {
                parts[i] = processRemainingText(parts[i]);
            }
        }
        return parts.join("");
    });
}

export function skipCodeBlocks(
    markdown: string,
    processRemainingText: (markdown: string) => string,
): string {
    const codeBlockRegex =
        /^(\s*)(`{3,})(.*?)[\r\n][\s\S]*?(?:\r|\n|\r\n)\1\2(?=$|[\r\n])/gm;
    const match = codeBlockRegex.exec(markdown);
    if (match) {
        return (
            processRemainingText(markdown.substring(0, match.index)) +
            match[0] +
            skipCodeBlocks(
                markdown.substring(match.index + match[0].length),
                processRemainingText,
            )
        );
    }
    return processRemainingText(markdown);
}

export class ObsidianUtils implements MediaCollector {
    private app: App;
    private fileSystem: FileSystemAdapter;
    private settings: SlidesExtendedSettings;
    private readonly processor: MarkdownProcessor;
    private readonly vaultDir;
    private readonly pluginDir;
    private readonly distDir;
    private readonly exportDir;
    private readonly cssSearchPath;
    private readonly highlightSearchPath;
    private readonly themeSearchPath;
    private readonly htmlTemplateSearchPath;

    private images = new Set<string>();
    private isCollecting = false;

    private yamlRegex = /^(---\n[\s\S]*?\n---\n)/;

    constructor(app: App, settings: SlidesExtendedSettings) {
        this.app = app;
        if (!(this.app.vault.adapter instanceof FileSystemAdapter)) {
            throw new Error("Slides Extended is only supported on the desktop");
        }

        this.fileSystem = this.app.vault.adapter;
        this.settings = settings;
        this.vaultDir = `${this.fileSystem.getBasePath()}/`;
        this.pluginDir = path.join(
            this.vaultDir,
            this.app.vault.configDir,
            "plugins/slides-extended/",
        );
        this.distDir = path.join(this.pluginDir, "dist/");
        this.exportDir = path.join(
            this.vaultDir,
            this.settings.exportDirectory,
        );

        this.cssSearchPath = [
            path.join(this.pluginDir), // relative to plugin dir
        ];
        this.highlightSearchPath = [
            path.join(this.pluginDir, "plugin/highlight"),
        ];
        this.themeSearchPath = [
            path.join(this.pluginDir, "css"), // plugin layouts
            path.join(this.distDir, "theme"), // reveal.js themes
        ];
        this.htmlTemplateSearchPath = [path.join(this.pluginDir, "template")];

        if (this.settings.themeDirectory) {
            const vaultThemeDir = path.join(
                this.vaultDir,
                this.settings.themeDirectory,
            );
            this.cssSearchPath.unshift(vaultThemeDir);
            this.highlightSearchPath.unshift(vaultThemeDir);
            this.themeSearchPath.unshift(vaultThemeDir);
            this.htmlTemplateSearchPath.unshift(`${vaultThemeDir}/html`);
        }

        setMediaCollector(this);
        this.processor = new MarkdownProcessor(this);
    }

    get markdownProcessor(): MarkdownProcessor {
        return this.processor;
    }

    get vaultName(): string {
        return this.fileSystem.getName();
    }
    get vaultDirectory(): string {
        return this.vaultDir;
    }

    get pluginDirectory(): string {
        return this.pluginDir;
    }

    get distDirectory(): string {
        return this.distDir;
    }

    get exportDirectory(): string {
        return this.exportDir;
    }

    getLocalCssSearchPath(): string[] {
        return this.cssSearchPath;
    }

    getThemeSearchPath(): string[] {
        return this.themeSearchPath;
    }
    getHighlightSearchPath(): string[] {
        return this.highlightSearchPath;
    }
    getHtmlTemplateSearchPath(): string[] {
        return this.htmlTemplateSearchPath;
    }

    getSettings(): SlidesExtendedSettings {
        return this.settings;
    }

    private getTFile(filename: string): TFile | null {
        if (filename.startsWith("[[") && filename.endsWith("]]")) {
            filename = filename.substring(2, filename.length - 2).trim();
        }
        filename = filename.replaceAll("%20", " ");

        const expDir = this.settings.exportDirectory.startsWith("/")
            ? this.settings.exportDirectory.substring(1)
            : this.settings.exportDirectory;

        const allFiles = this.app.vault.getFiles();
        const filesNotInExportDir = allFiles.filter(
            (item) => !item.path.contains(expDir),
        );
        const allHits = filesNotInExportDir.filter((item) =>
            item.path.contains(filename),
        );

        let file: TFile = null;

        if (allHits.length === 1) {
            // Only one match
            file = allHits.first();
        }

        if (!file && filename.toLowerCase().endsWith(".excalidraw")) {
            // Workaround for Excalidraw images
            let hit = filesNotInExportDir.filter((x) =>
                x.path.contains(`${filename}.svg`),
            );
            if (hit) {
                file = hit.first();
            } else {
                hit = filesNotInExportDir.filter((x) =>
                    x.path.contains(`${filename}.png`),
                );
                if (hit) {
                    file = hit.first();
                }
            }
        }

        if (!file && allHits.length > 1) {
            // Find file most similar to search term
            let score = 0;
            for (const hit of allHits) {
                const currentScore = this.similarity(filename, hit.path);
                if (currentScore > score) {
                    score = currentScore;
                    file = hit;
                }
            }
        }

        return file;
    }

    getAbsolutePath(relativePath: string): string {
        const markdownFile = this.getTFile(relativePath);
        return this.absolute(markdownFile?.path);
    }

    getRelativePath(path: string): string | null {
        if (path == null) {
            return null;
        }
        const file = this.getTFile(path);
        return file?.path;
    }

    absolute(relativePath: string) {
        return relativePath ? this.fileSystem.getFullPath(relativePath) : null;
    }

    findFile(path: string) {
        const file: TFile = this.getTFile(path);
        console.debug("findFile", path, file);
        return file ? file.path : path;
    }

    findMediaFile(path: string) {
        let base = "";
        if (!getMediaCollector().shouldCollect()) {
            base = "/";
        }
        const file: TFile = this.getTFile(path);
        console.debug("findMediaFile", path, file, base);
        return file ? base + file.path : path;
    }

    parseFile(filename: string, header: string) {
        const tfile = this.getTFile(filename);

        if (!tfile) {
            return null;
        }

        const absoluteFilePath = this.absolute(tfile?.path);
        const fileContent = readFileSync(absoluteFilePath, {
            encoding: "utf-8",
        });

        if (header === null) {
            const match = this.yamlRegex.exec(fileContent);
            if (match) {
                return fileContent.slice(match[0].length);
            }
            return fileContent;
        }
        const cache = this.app.metadataCache.getFileCache(tfile);
        const resolved = resolveSubpath(cache, header);
        console.log("parseFile, cache, resolved", cache, resolved);

        if (resolved && resolved.start && resolved.start.line != null) {
            if (resolved.end && resolved.end.line != null) {
                return this.substring(
                    fileContent,
                    resolved.start.line,
                    resolved.start.col,
                    resolved.end.line,
                    resolved.end.col,
                );
            }
            return this.substring(
                fileContent,
                resolved.start.line,
                resolved.start.col,
                -1,
                -1,
            );
        }
        return `![[${filename}#${header}]]`;
    }

    async fetchRemoteMarkdown(markdown: string): Promise<string> {
        const wikilinkFileRegex = /!\[\[(file:.+?\.md)(\|[^\]]+)?\]\]/gi;
        const fileUrlRegex = /(!\[[^\]]*?\]\()(file:.+?\.md(?:#.+?)?)(\))/i;

        // Replace wikilinks with markdown links
        markdown = markdown.replace(wikilinkFileRegex, (_match, p1, p2) => {
            const filePath = p1;
            const alias = p2 ? p2.slice(1) : "";
            const url = new URL(filePath);
            return `![${alias}](${url})`;
        });

        // Replace markdown links with markdown content
        if (fileUrlRegex.test(markdown)) {
            const lines = markdown.split("\n");
            for (let index = 0; index < lines.length; index++) {
                const line = lines[index];
                const match = fileUrlRegex.exec(line);
                // one per line (as embed, should be fine)
                if (match) {
                    const fullMatch = match[0];
                    const fileUrl = match[2];
                    const content = await this.readRemoteFile(fileUrl);
                    if (content) {
                        lines[index] = line.replace(fullMatch, content);
                    }
                }
            }
            markdown = lines.join("\n");
        }

        return markdown;
    }

    async readRemoteFile(fileUrl: string): Promise<string> {
        const anchorIndex = fileUrl.indexOf("#");
        if (anchorIndex > -1) {
            fileUrl = fileUrl.substring(0, anchorIndex);
        }
        const urlpath = fileUrl.replace("file://", Platform.resourcePathPrefix);

        const result = await fetch(urlpath).catch((error) => {
            return new Response(null, {
                status: 404,
                statusText: error.messge,
            });
        });
        if (result.ok) {
            if (result.blob) {
                const blob = await result.blob();
                const bytes = await blob.arrayBuffer();
                const content = Buffer.from(bytes).toString();
                if (this.yamlRegex.test(content)) {
                    return this.yamlRegex.exec(content)[1];
                }
                return content;
            }
            console.info(
                "open a bug to handle this kind of response. Include this message",
                result,
            );
        }
        return null;
    }

    substring(
        input: string,
        startLine: number,
        startColumn: number,
        endLine: number,
        endColumn: number,
    ): string {
        let result = "";
        const lines = input.split("\n");

        let eline = lines.length;
        if (endLine > -1) {
            eline = endLine;
        }

        for (let index = startLine; index <= eline; index++) {
            const line = lines[index];
            if (line != null) {
                if (index === startLine) {
                    result += `${line.substring(startColumn)}\n`;
                } else if (index === eline) {
                    let endLine = line;
                    if (endColumn > -1) {
                        endLine = line.substring(0, endColumn);
                    }
                    if (endLine.includes("^")) {
                        endLine = endLine.substring(
                            0,
                            endLine.lastIndexOf("^"),
                        );
                    }
                    result += `${endLine}\n`;
                } else {
                    result += `${line}\n`;
                }
            }
        }

        return result;
    }

    similarity(s1: string, s2: string): number {
        let longer = s1;
        let shorter = s2;
        if (s1.length < s2.length) {
            longer = s2;
            shorter = s1;
        }
        const longerLength = longer.length;
        if (longerLength === 0) {
            return 1.0;
        }
        return (
            (longerLength - this.editDistance(longer, shorter)) / longerLength
        );
    }

    editDistance(s1: string, s2: string): number {
        s1 = s1.toLowerCase();
        s2 = s2.toLowerCase();

        const costs = [];
        for (let i = 0; i <= s1.length; i++) {
            let lastValue = i;
            for (let j = 0; j <= s2.length; j++) {
                if (i === 0) costs[j] = j;
                else {
                    if (j > 0) {
                        let newValue = costs[j - 1];
                        if (s1.charAt(i - 1) !== s2.charAt(j - 1))
                            newValue =
                                Math.min(
                                    Math.min(newValue, lastValue),
                                    costs[j],
                                ) + 1;
                        costs[j - 1] = lastValue;
                        lastValue = newValue;
                    }
                }
            }
            if (i > 0) costs[s2.length] = lastValue;
        }
        return costs[s2.length];
    }

    resetImageCollection() {
        console.debug(
            "enable image collection for exported slides, start collecting",
        );
        this.images.clear();
        this.isCollecting = true;
    }

    disableImageCollection() {
        console.debug("stop collecting images", this.images);
        this.isCollecting = false;
    }

    // MediaCollector
    addMedia(filePath: string): void {
        this.images.add(filePath);
    }

    getAll(): string[] {
        return Array.of(...this.images);
    }

    shouldCollect(): boolean {
        return this.isCollecting;
    }
}
