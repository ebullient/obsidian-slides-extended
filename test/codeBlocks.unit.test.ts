import { MarkdownProcessor } from 'src/obsidian/markdownProcessor';
import { obsidianUtils as utilsInstance } from './__mocks__/mockObsidianUtils';
import { prepare } from './testUtils';
import { readFileSync } from 'node:fs';

test('Code Block Syntax > Code Blocks', () => {
	const input = `
\`\`\`dockerfile
FROM ubuntu
\`\`\`
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > Code Blocks with $ and underscores', () => {
	const input = `
\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > no type', () => {
	const input = `
\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > codeblock-ish, not math', () => {
	const input = `
 \`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks.
Underscores should not be escaped
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > Math with Code Blocks', () => {
	const input = readFileSync('test/fixtures/mathjax-codeblock.md', 'utf8');

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > Math with Mixed Code Blocks', () => {
	const input = readFileSync('test/fixtures/mathjax-codeblock-mixed.md', 'utf8');

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Embedded code has extra characters near dollar signs', () => {
	const input = readFileSync('test/fixtures/codeblock-with-math-chars.md', 'utf8');

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

