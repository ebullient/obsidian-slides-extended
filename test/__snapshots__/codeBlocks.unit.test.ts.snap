// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Code Block Syntax > Code Blocks 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`dockerfile
FROM ubuntu
\`\`\`
</div>"
`;

exports[`Code Block Syntax > Code Blocks with $ and underscores 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`
</div>"
`;

exports[`Code Block Syntax > Math with Code Blocks 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

$$\\begin{vmatrix}a & b\\\\\\c & d\\end{vmatrix}=ad-bc$$

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`

You can also do inline math like $s^{-2}\\_{n}\\sum\\_{i=1}^{n}$
</div>"
`;

exports[`Code Block Syntax > Math with Mixed Code Blocks 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

$$\\begin{vmatrix}a & b\\\\\\c & d\\end{vmatrix}=ad-bc$$

\`\`\`\`md
\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`
\`\`\`\`

You can also use inline syntax (\`$s^{-2}_{n}\\sum_{i=1}^{n}$\`)
to do inline math like $s^{-2}\\_{n}\\sum\\_{i=1}^{n}$
</div>"
`;

exports[`Code Block Syntax > codeblock-ish, not math 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks.
Underscores should not be escaped
</div>"
`;

exports[`Code Block Syntax > no type 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks
</div>"
`;

exports[`Embedded code has extra characters near dollar signs 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`javascript
export class CounterService {
  count$ = new BehaviorSubject(1000);

  double$ = this.count$.pipe(map((count) => count * 2));
  triple$ = this.count$.pipe(map((count) => count * 3));

  combined$ = combineLatest([this.double$, this.triple$]).pipe(
    map(([double, triple]) => double + triple)
  );

  over9000$ = this.combined$.pipe(map((combined) => combined > 9000));

  message$ = this.over9000$.pipe(
    map((over9000) => (over9000 ? "It's over 9000!" : "It's under 9000."))
  );
}
\`\`\`
</div>"
`;
